File: examples/ReactNative/TimeKeeper+LinkMobile/ios/BridgeModule.swift
import Foundation
import LinkMobile
import React

@objc(BridgeModule)
class BridgeModule: RCTEventEmitter {
    private var seamlessBridge: Bridge?
    private var blueLinkListener: BlueLinkListenerBridge?

    override static func requiresMainQueueSetup() -> Bool {
        return true
    }

    override func supportedEvents() -> [String]! {
        // List of all possible event names that JS can listen to
        return [
            "onDeviceDiscovered",
            "onDeviceConnected",
            "onDeviceDisconnected",
            "onBLEStateChanged",
        ]
    }

    @objc
    func initialize(
        _ serverAddress: String,
        uuid: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        DispatchQueue.main.async {
            self.seamlessBridge = Bridge(serverAddress: serverAddress, uuid: uuid)
            self.seamlessBridge?.setLogsEnabled(isEnabled: true)
            guard let blueLink = self.seamlessBridge?.getBlueLink() else {
                rejecter("ERROR", "Failed to get BlueLink from Bridge", nil)
                return
            }

            // Attach listener bridge
            self.blueLinkListener = BlueLinkListenerBridge(eventEmitter: self)
            blueLink.addListener(listener: self.blueLinkListener!, prependListener: false)

            resolver("Bridge initialized and listener attached")
        }
    }

    @objc
    func start(
        _ uuid: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        guard let seamlessBridge = self.seamlessBridge else {
            rejecter("ERROR", "Bridge not initialized. Call initialize first.", nil)
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            seamlessBridge.start(UUID: uuid)
            DispatchQueue.main.async {
                resolver("Bridge started with UUID: \(uuid)")
            }
        }
    }

    // Required to avoid warnings
    override func startObserving() {}
    override func stopObserving() {}
}

class BlueLinkListenerBridge: BlueLinkEventListener {
    weak var eventEmitter: RCTEventEmitter?

    init(eventEmitter: RCTEventEmitter) {
        self.eventEmitter = eventEmitter
        super.init()
    }

    override func onBLEStateChanged(newState: Bool) {
        eventEmitter?.sendEvent(
            withName: "onBLEStateChanged",
            body: [
                "state": newState
            ])
    }

    override func onDeviceDiscovered(
        device: BluetoothDevice,
        advertisementData: BluetoothAdvertisement,
        rssi: Int32,
        signalStrength: SignalStrength?,
        connectionStatus: BlueLinkDeviceConnectionStatus
    ) {
        print("[BlueLinkListenerBridge] onDeviceDiscovered called")

        // Send simple serializable info to JS
        eventEmitter?.sendEvent(
            withName: "onDeviceDiscovered",
            body: [
                "name": device.name ?? "Unknown",
                "identifier": device.deviceId,
                "rssi": rssi,
                "connectionStatus": connectionStatus.name,
                "signalStrength": signalStrength?.description ?? "unknown",
            ])
    }

    override func onDeviceConnected(device: BluetoothDevice) {
        eventEmitter?.sendEvent(
            withName: "onDeviceConnected",
            body: [
                "name": device.name ?? "Unknown",
                "identifier": device.deviceId,
            ])
    }

    override func onDeviceDisconnected(device: BluetoothDevice, reason _: KotlinInt?) {
        eventEmitter?.sendEvent(
            withName: "onDeviceDisconnected",
            body: [
                "name": device.name ?? "Unknown",
                "identifier": device.deviceId,
            ])
    }
}

File: examples/ReactNative/TimeKeeper+LinkMobile/android/app/src/main/java/com/cody669/TimeKeeper/BridgeModule.kt
package com.cody669.TimeKeeper

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.lastlock.bridge.Bridge
import android.util.Log
import com.lastlock.bridge.bluelink.BluetoothAdvertisement
import com.lastlock.bridge.bluelink.BluetoothDevice
import com.lastlock.bridge.bluelink.BlueLinkDeviceConnectionStatus
import com.lastlock.bridge.bluelink.interfaces.BlueLinkEventListener
import com.lastlock.bridge.bluelink.interfaces.SignalStrength

class BridgeModule(reactContext: ReactApplicationContext) :
        ReactContextBaseJavaModule(reactContext) {

  override fun getName() = "BridgeModule"

  private fun sendEvent(eventName: String, params: WritableMap?) {
    reactApplicationContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit(eventName, params)
  }

  @ReactMethod
  fun initialize(serverAddress: String, uuid: String, promise: Promise) {
    try {
      Bridge.init(reactApplicationContext, serverAddress)
      Bridge.setLogsEnabled(true)

      // Attach listener to receive device discovery events
      val blueLink = Bridge.getBlueLink()
      blueLink?.addListener(object : BlueLinkEventListener() {
        override fun onDeviceDiscovered(
          device: BluetoothDevice,
          advertisementData: BluetoothAdvertisement,
          rssi: Int,
          signalStrength: SignalStrength,
          connectionStatus: BlueLinkDeviceConnectionStatus
        ) {
          Log.d("BridgeModule", "onDeviceDiscovered called! deviceId=${device.deviceId.toString()}")

          val params = Arguments.createMap().apply {
            putString("name", device.name ?: "Unknown")
            putString("identifier", device.deviceId.toString())
            putInt("rssi", rssi)
            putString("connectionStatus", connectionStatus.name)
            putString("signalStrength", signalStrength.toString() ?: "unknown")
          }
          sendEvent("onDeviceDiscovered", params)
        }

        override fun onDeviceConnected(device: BluetoothDevice) {
          val params = Arguments.createMap().apply {
            putString("name", device.name ?: "Unknown")
            putString("identifier", device.deviceId.toString())
          }
          sendEvent("onDeviceConnected", params)
        }

        override fun onDeviceDisconnected(device: BluetoothDevice, reason: Int?) {
          val params = Arguments.createMap().apply {
            putString("name", device.name ?: "Unknown")
            putString("identifier", device.deviceId.toString())
          }
          sendEvent("onDeviceDisconnected", params)
        }

        override fun onBLEStateChanged(newState: Boolean) {
          val params = Arguments.createMap().apply {
            putBoolean("state", newState)
          }
          sendEvent("onBLEStateChanged", params)
        }
      }, false)

      promise.resolve("Bridge initialized")
    } catch (e: Exception) {
      promise.reject("INIT_ERROR", e)
    }
  }

  @ReactMethod
  fun start(uuid: String, promise: Promise) {
    try {
      Bridge.start(uuid)
      promise.resolve("Bridge started with UUID: $uuid")
    } catch (e: Exception) {
      promise.reject("START_ERROR", e)
    }
  }

  // Required for NativeEventEmitter
  @ReactMethod
  fun addListener(eventName: String) {
    // Keep: Required for RN built-in Event Emitter Calls
  }

  @ReactMethod
  fun removeListeners(count: Int) {
    // Keep: Required for RN built-in Event Emitter Calls
  }
}